module: support eval with ts syntax detection (#56285)
nodejs/node: doc/api/cli.md

-->

This configures Node.js to interpret `--eval` or `STDIN` input as CommonJS or
as an ES module. Valid values are `"commonjs"`, `"module"`, `"module-typescript"` and `"commonjs-typescript"`.
The `"-typescript"` values are available only in combination with the flag `--experimental-strip-types`.
The default is `"commonjs"`.

If `--experimental-strip-types` is enabled and `--input-type` is not provided,
Node.js will try to detect the syntax with the following steps:

1. Run the input as CommonJS.
2. If step 1 fails, run the input as an ES module.

> 3. If step 2 fails with a SyntaxError, strip the types.

Reviewer: Any syntax error?

Author: Yes, we can narrow down the scope when we find edge cases

-->

This configures Node.js to interpret `--eval` or `STDIN` input as CommonJS or
as an ES module. Valid values are `"commonjs"`, `"module"`, `"module-typescript"` and `"commonjs-typescript"`.
The `"-typescript"` values are available only in combination with the flag `--experimental-strip-types`.
The default is `"commonjs"`.

If `--input-type`is not provided, Node.js will try to run the code as JavaScript,
then it will try to run the code as TypeScript.

> When using TypeScript, the `--input-type=type` flag should be used,

Author: ```suggestion
When using TypeScript, the `--input-type=type` flag should always be used,
```

> to reduce the overhead of multiple syntax detection.

Author: ```suggestion
to reduce the overhead of multiple syntax detections.
```

-->

This configures Node.js to interpret `--eval` or `STDIN` input as CommonJS or
as an ES module. Valid values are `"commonjs"`, `"module"`, `"module-typescript"` and `"commonjs-typescript"`.

> The `"-typescript"` values are available only in combination with the flag `--experimental-strip-types`.

Reviewer: Just confirming, I assume that once we unflag strip-types then all that happens is that this sentence disappears? We don’t create some other new flag that needs to be passed to permit these new values?

Author: Exactly no need for additional flags

The default is `"commonjs"`.

> If `--experimental-strip-types` is enabled and `--input-type`is not provided,

Reviewer: ```suggestion
If `--experimental-strip-types` is enabled and `--input-type` is not provided,
```

Node.js will first try to run the code as JavaScript,

> then it will try to run the code as TypeScript.

Reviewer: We should probably say something about syntax detection here, that first it attempts to run as CommonJS JavaScript, and depending on the parse error thrown, will retry as ESM JavaScript or CommonJS TypeScript; and then potentially a third time as ESM TypeScript? I’m not even sure what the flow would be, which probably means that we should spell it out 😄

Author: I'm not an expert on how the syntax detection is implemented but the flow is:
- run normally as javascript (this will perform the detection by trying cjs and esm)
- if invalid syntax, strip the types and try again
- if invalid again throw the first invalid syntax error enriched with the typescript message

Reviewer: The details of syntax detection are here: https://github.com/nodejs/node/blob/6012a4e9a2733b25a7023baa0a58bef85c9ebc5f/doc/api/esm.md#:~:text=at%20pjsonURL.-,DETECT_MODULE_SYNTAX,-(source)

The important part is to clarify which specific errors will trigger strip-types behavior. I assume there’s no overlap between the “retry as ESM” errors and the “retry with type stripping” errors, so the flow is no more expensive than the current detection flow: try first as CommonJS JavaScript, and retry as stripped types.

It gets even more complicated because doesn’t the strip-types flow itself use syntax detection? Like after it strips the types, then it initially tries to run as CommonJS, then retries as ESM?

Author: Yes

Reviewer: I think the implementation is fine, let’s just improve the documentation to better explain it? Is it this?

1. Try to run as CommonJS JavaScript.
2. If the above fails because of a parse error related to module syntax, retry as ESM JavaScript.
3. If either of the above fails because of a parse error related to types syntax, strip the types and run as stripped CommonJS JavaScript.
4. If the above fails because of a parse error related to module syntax, retry as ESM JavaScript.

> When using TypeScript, the `--input-type=type` flag should always be used,
> to reduce the overhead of multiple syntax detections.

Reviewer: This is a bit strong. I don’t think people using `--eval` are generally concerned about performance, so parsing a <1 KB string multiple times isn’t a big deal. Rather than making a strong recommendation, we can just offer a hint on how to avoid the performance hit for those who care:

```suggestion
To avoid the delay of multiple syntax detection passes, the `--input-type=type` flag can be used to specify
how the `--eval` input should be interpreted.
```

-->

This configures Node.js to interpret `--eval` or `STDIN` input as CommonJS or
as an ES module. Valid values are `"commonjs"`, `"module"`, `"module-typescript"` and `"commonjs-typescript"`.
The `"-typescript"` values are available only in combination with the flag `--experimental-strip-types`.
The default is `"commonjs"`.

If `--experimental-strip-types` is enabled and `--input-type` is not provided,
Node.js will try to detect the syntax with the following steps:

1. Run the input as CommonJS.
2. If step 1 fails, run the input as an ES module.
3. If step 2 fails with a SyntaxError, strip the types.
4. If step 3 fails with an error code [`ERR_INVALID_TYPESCRIPT_SYNTAX`][],
   throw the error from step 2, including the TypeScript error in the message,
   else run as CommonJS.
5. If step 4 fails, run the input as an ES module.

> To avoid the delay of multiple syntax detection passes, the `--input-type=type` flag can be used to specify

Reviewer 2: I think with the current way the code is, the delay from syntax detection is not the biggest caveat - it's the fact that the code would be executed twice...(e.g. if you `console.log`, you'll see two logs; if you send a network request, it will be sent twice, etc.)
